### ISSUE TITLE
OpenLDAP groups not mapped and performances

### ISSUE THREAD

**Issue by ghost** [nextcloud/server#456]
### Steps to reproduce
1. Install NextCloud 9.0.51.0
2. Configure LDAP authentication against an OpenLDAP directory which implements groups using MemberUid
### Expected behaviour

Users groups in NC should correspond to LDAP groups.
**It works fine with OwnCloud 8 on the same server with same LDAP configuration**.
### Actual behaviour
- users don't have any group
- group search is very slow (for example in completion when selecting groups for external storage permissions)
### Server configuration

**Operating system**:
Ubuntu 16.04

**Web server:**
Apache 2.4.18

**Database:**
MariaDB 10.0.25

**PHP version:** 
7.0.4

**Nextcloud version:** 
9.0.51.0

**Updated from an older Nextcloud/ownCloud or fresh install:**
no, fresh install

**Where did you install Nextcloud from:** 
Manual Installation on Linux

**Signing status:** 
<details>
<summary>

Signing status</summary>

```
No errors have been found
```

</details>

**List of activated apps:**
<details>
<summary>

App list</summary>

```
Enabled:
  - admin_audit: 1.0.0
  - comments: 0.2
  - dav: 0.1.6
  - documents: 0.12.0
  - federatedfilesharing: 0.1.0
  - files: 1.4.4
  - files_external: 0.5.2
  - files_pdfviewer: 0.8.1
  - files_sharing: 0.9.1
  - files_texteditor: 2.1
  - files_trashbin: 0.8.0
  - files_versions: 1.2.0
  - provisioning_api: 0.4.1
  - systemtags: 0.2
  - updatenotification: 0.1.0
  - user_ldap: 0.8.0
Disabled:
  - activity
  - encryption
  - external
  - federation
  - files_videoplayer
  - firstrunwizard
  - gallery
  - libreonline
  - notifications
  - templateeditor
  - user_external
```

</details>

**The content of config/config.php:**
<details>
<summary>

Config report</summary>

```
{
    "system": {
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "localhost",
            "172.31.0.42",
            "owncloud.cn.isf"
        ],
        "datadirectory": "\/var\/www\/nextcloud\/data",
        "overwrite.cli.url": "http:\/\/localhost",
        "dbtype": "mysql",
        "version": "9.0.51.0",
        "dbname": "nextcloud",
        "dbhost": "localhost",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "logtimezone": "UTC",
        "installed": true,
        "instanceid": "oc2nh0djafzp",
        "appstore.experimental.enabled": true,
        "ldapIgnoreNamingRules": false
    }
}
```

</details>

**Are you using external storage, if yes which one:**
yes, local

**Are you using encryption:** 
no

**Are you using an external user-backend, if yes which one:** 
LDAP
#### LDAP configuration (delete this part if not used)

<details>
<summary>

LDAP config</summary>

```
+-------------------------------+------------------------------------------------------------------------------------------+
| Configuration                 | s01                                                                                      |
+-------------------------------+------------------------------------------------------------------------------------------+
| hasMemberOfFilterSupport      | 1                                                                                        |
| hasPagedResultSupport         |                                                                                          |
| homeFolderNamingRule          |                                                                                          |
| lastJpegPhotoLookup           | 0                                                                                        |
| ldapAgentName                 | cn=foo,dc=bar,dc=org                                                          |
| ldapAgentPassword             | ***                                                                                      |
| ldapAttributesForGroupSearch  | cn                                                                                       |
| ldapAttributesForUserSearch   |                                                                                          |
| ldapBackupHost                |                                                                                          |
| ldapBackupPort                |                                                                                          |
| ldapBase                      | ou=members,ou=people,dc=bar,dc=org                                           |
| ldapBaseGroups                | ou=groups,ou=groups,dc=bar,dc=org                                    |
| ldapBaseUsers                 | ou=members,ou=people,dc=bar,dc=org                                           |
| ldapCacheTTL                  | 600                                                                                      |
| ldapConfigurationActive       | 1                                                                                        |
| ldapDynamicGroupMemberURL     |                                                                                          |
| ldapEmailAttribute            | mail                                                                                     |
| ldapExperiencedAdmin          | 0                                                                                        |
| ldapExpertUUIDGroupAttr       |                                                                                          |
| ldapExpertUUIDUserAttr        |                                                                                          |
| ldapExpertUsernameAttr        |                                                                                          |
| ldapGroupDisplayName          | cn                                                                                       |
| ldapGroupFilter               | (&(|(objectclass=posixGroup)))                                                           |
| ldapGroupFilterGroups         |                                                                                          |
| ldapGroupFilterMode           | 1                                                                                        |
| ldapGroupFilterObjectclass    | top                                                                                      |
| ldapGroupMemberAssocAttr      | memberUid                                                                                |
| ldapHost                      | ldaphost.bar.org                                                                             |
| ldapIgnoreNamingRules         |                                                                                          |
| ldapLoginFilter               | (&(&(|(objectclass=inetOrgPerson)))(|(uid=%uid)(|(mailPrimaryAddress=%uid)(mail=%uid)))) |
| ldapLoginFilterAttributes     |                                                                                          |
| ldapLoginFilterEmail          | 1                                                                                        |
| ldapLoginFilterMode           | 0                                                                                        |
| ldapLoginFilterUsername       | 1                                                                                        |
| ldapNestedGroups              | 0                                                                                        |
| ldapOverrideMainServer        |                                                                                          |
| ldapPagingSize                | 500                                                                                      |
| ldapPort                      | 389                                                                                      |
| ldapQuotaAttribute            |                                                                                          |
| ldapQuotaDefault              |                                                                                          |
| ldapTLS                       | 0                                                                                        |
| ldapUserDisplayName           | uid                                                                                      |
| ldapUserDisplayName2          | displayname                                                                              |
| ldapUserFilter                | (&(|(objectclass=inetOrgPerson)))                                                        |
| ldapUserFilterGroups          |                                                                                          |
| ldapUserFilterMode            | 0                                                                                        |
| ldapUserFilterObjectclass     | inetOrgPerson                                                                            |
| ldapUuidGroupAttribute        | auto                                                                                     |
| ldapUuidUserAttribute         | auto                                                                                     |
| turnOffCertCheck              | 0                                                                                        |
| useMemberOfToDetectMembership | 1                                                                                        |
+-------------------------------+------------------------------------------------------------------------------------------+
```

</details>
### Client configuration

**Browser:** 
all

**Operating system:** 
all
#### Web server error log

<details>
<summary>

Web server error log</summary>

```
Insert your webserver log here
```

</details>
#### Nextcloud log (data/nextcloud.log)

<details>
<summary>

Nextcloud log</summary>

No relevant errors.
</details>
#### Browser log

<details>
<summary>

Browser log</summary>

```
No relevant logs.
```

</details>

Regards,
Benjamin

**Comment by LukasReschke** [comment]
cc @blizzz

**Comment by bigbrovar** [comment]
Same issue here. I also noticed the ldap user mail attribute is not mapped.

**Comment by blizzz** [comment]
> Users groups in NC should correspond to LDAP groups.
It works fine with OwnCloud 8 on the same server with same LDAP configuration.

*Should* work just as well ;)

@ghost can you set log level to debug and provide the log output when running ``./occ ldap:search --group $name`` (replace ``$name`` accordingly)?

> Same issue here. I also noticed the ldap user mail attribute is not mapped.

@bigbrovar is the attribute specified in the settings?

**Comment by kfeine** [comment]
@blizzz Hi, here are the logs you asked when I search 'mygroup':

    Dec  9 19:34:07 owncloud ownCloud[9072]: {user_ldap} getGroups getGroups-mygroup-15-0
    Dec  9 19:34:07 owncloud ownCloud[9072]: {user_ldap} getGroups Filter (&(&(|(objectclass=posixGroup)))(cn=mygroup*))
    Dec  9 19:34:07 owncloud ownCloud[9072]: {user_ldap} initializing paged search for  Filter (&(&(|(objectclass=posixGroup)))(cn=mygroup*)) base Array#012(#012    [0] => ou=member,ou=memberships,dc=domain,dc=com#012)#012 attr Array#012(#012    [0] => cn#012    [1] => dn#012)#012 limit 15 offset 0
    Dec  9 19:34:07 owncloud ownCloud[9072]: {user_ldap} Ready for a paged search
    Dec  9 19:34:07 owncloud ownCloud[9072]: {user_ldap} getGroups(mygroup): read 1 at offset 0 (limit: 15)

Regards,

**Comment by blizzz** [comment]
@kfeine you are/were @ghost I suppose?

The log looks OK, and I assume the group was also printed out on the command line?

> users don't have any group

did you recheck that "memberUid" as association attribute is correct?

> group search is very slow (for example in completion when selecting groups for external storage permissions)

how do you quantify "very slow"? Is it only so in files_external, or also in other places (e.g. sharing)?

**Comment by kfeine** [comment]
Hi,
Sorry for the late reply, I've been away for a while.

> did you recheck that "memberUid" as association attribute is correct?

I'm sorry I don't get what you mean by that…

Anyway, I upgraded to NC 11 to see if it solved the problem, but the same behavior occurred. I tried to reverse engineer the issue, and I found that the if-block lines 500 to 520 in ``apps/user_ldap/lib/Group_LDAP.php`` are the cause of this issue: it seems to me the condition is always true, so the next part with ``memberuid`` search is never called. If I comment this block it runs as expected.

Thanks,

**Comment by blizzz** [comment]
@kfeine this one? https://github.com/nextcloud/server/blob/stable11/apps/user_ldap/lib/Group_LDAP.php#L499

If this turns out to be true, then 1) it is indicated that the memberof-overlay is enabled on you LDAP Server, i.e. specyfying ``memberof=$GROUPDN`` would work in a filter. If, for some reason, it does not work with your setup, you can turn it off by setting ``useMemberOfToDetectMembership`` to  0 via cli:

```sudo -u $WEBUSER ./occ ldap:set-config $CONFIGID useMemberOfToDetectMembership 0```

**Comment by andreas-p** [comment]
user_ldap should check if the memberof attribute is identical to the group membership attribute, and disable use_memberof_to_detect_membership by default if they're different attributes.

**Comment by blizzz** [comment]
It is checked whether ``memberOf=`` returns any results and only enabled when that is the case.

**Comment by andreas-p** [comment]
if group membership is checked against posix groups, memberof cannot work because it doesn't support memberUid. Only member and uniqueMember are supported.

**Comment by kfeine** [comment]
@blizzz you're right, we had a memberof-overlay enabled on our ldap test server… setting useMemberOfToDetectMembership to 0 solved the group mapping issue.

Sorry for the issue, and thank you very much for you work on this product!

**Comment by andreas-p** [comment]
See https://github.com/nextcloud/server/pull/3329

**Comment by blizzz** [comment]
@kfeine glad it's solved for you, I close this issue then.

@andreas-p thanks for the PR, I will have a look with the next time slice.

**Comment by andreas-p** [comment]
The minimalistic fix https://github.com/nextcloud/server/pull/3329 isn't reviewed/committed nearly 2 months later. As long as an issue isn't fixed in git, it shouldn't be closed.

**Comment by blizzz** [comment]
😱 Ooops, added my review.

**Comment by rullzer** [comment]
Fixed with#3329
